src/modules/storage/BlockStorage.ts

```diff
@@ -0,0 +1,152 @@
+            sqlite.OPEN_SHAREDCACHE);
+        this.db.configure("busyTimeout", 1000);
+        this.createTable(callback);
+    }
+
+    /**
+     * Creates a table. table name is `block`.
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    public createTable (callback : any = null)
+    {
+        var sql =
+        `CREATE TABLE IF NOT EXISTS blocks
+        (
+            height INTEGER PRIMARY KEY,
+            prev_block TEXT,
+            validators TEXT,
+            merkle_root TEXT,
+            signature TEXT
+        )`;
+        this.db.run(sql, (err : any) =>
+        {
+            if (callback != null)
+                callback(err);
+        });
+    }
+
+    /**
+     * Puts a block to database
+     * @param data a block data
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    public put (data : any, callback : any = null)
+    {
+        if (
+                (data == null) ||
+                (data.header == undefined) ||
+                (data.header.height == undefined) ||
+                (data.header.height.value == undefined) ||
+                (data.header.prev_block == undefined) ||
+                (data.header.validators == undefined) ||
+                (data.header.validators._storage == undefined) ||
+                (data.header.merkle_root == undefined) ||
+                (data.header.signature == undefined)
+        ) {
+            if (callback != null)
+                callback("Parameter validation failed.");
+            return;
+        }
+
+        var sql =
+        `INSERT INTO blocks
+            (height, prev_block, validators, merkle_root, signature)
+        VALUES
+            (?, ?, ?, ?, ?)
+        ON CONFLICT(height) DO UPDATE SET
+            prev_block = prev_block,
+            validators = validators,
+            merkle_root = merkle_root,
+            signature = signature`;
+        this.db.run(sql,
+            [
+                data.header.height.value,
+                data.header.prev_block,
+                JSON.stringify(data.header.validators._storage),
+                data.header.merkle_root,
+                data.header.signature
+            ], (err : any) =>
+        {
+            if (callback != null)
+                callback(err);
+        });
+    }
+
+    /**
+     * Gets a block data
+     * @param height the height of the block to get
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     * The second argument is result set.
+     */
+    public get (height : number, callback : any = null)
```

Comment:
The callback seems to have a defined type, doesn't it ?

Reply:
If there is a value to be returned, the second parameter is valid.

Reply:
That doesn't relate to the question